### PR TITLE
SelectFwe added issue #56

### PR DESCRIPTION
**Pull request by kadarakos**
Solving issue:  Add SelectFwe to feature selection operators #56. 
Code is based on the SelectKBest operator.

### CODE REVIEW

File: .idea/.name [hunk 1]
```diff
@@ -0,0 +1 @@
+tpot
```
Comment (rhiever) [inline .idea/.name]: Please remove all of these files listed in the `.idea` directory.

File: tpot/tpot.py [hunk 2]
```diff
@@ -348,8 +349,9 @@ def export(self, output_file_name):
 '''
         if '_variance_threshold' in operators_used: pipeline_text += 'from sklearn.feature_selection import VarianceThreshold\n'
         if '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectKBest\n'
+        if '_select_fwe' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectFwe\n' # added selectfwe
         if '_select_percentile' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectPercentile\n'
-        if '_select_percentile' in operators_used or '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import f_classif\n'
+        if '_select_percentile' in operators_used or '_select_kbest' or '_select_fwe' in operators_used: pipeline_text += 'from sklearn.feature_selection import f_classif\n' # added selectfwe
```
Comment (rhiever) [inline tpot/tpot.py]: Please change this line to

```python
if '_select_percentile' in operators_used or '_select_kbest' in operators_used or '_select_fwe' in operators_used: pipeline_text += 'from sklearn.feature_selection import f_classif\n'
```

File: tpot/tpot.py [hunk 3]
```diff
@@ -738,6 +740,50 @@ def _select_kbest(self, input_df, k):
         mask_cols = list(training_features.iloc[:, mask].columns) + ['guess', 'class', 'group']
         return input_df[mask_cols].copy()
 
+
+    def _select_fwe(self, input_df, alpha):
+        """ BASED ON THE _select_kbest METHOD
```
Comment (rhiever) [inline tpot/tpot.py]: No need to include this line in the comments. :-)

File: tpot/tpot.py [hunk 4]
```diff
@@ -738,6 +740,50 @@ def _select_kbest(self, input_df, k):
...
+
+        Returns
+        -------
+        subsetted_df: pandas.DataFrame {n_samples, n_filtered_features + ['guess', 'group', 'class']}
+            Returns a DataFrame containing the 'best' features
+
+        """
+        training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1)
+        training_class_vals = input_df.loc[input_df['group'] == 'training', 'class'].values
+        alpha = float(alpha) # force float
```
Comment (rhiever) [inline tpot/tpot.py]: No need to force it to a float. We document it as requiring a float, so if the user passes an invalid value, it should error out on its own.

File: tpot/tpot.py [hunk 5]
```diff
@@ -738,6 +740,50 @@ def _select_kbest(self, input_df, k):
...
+    def _select_fwe(self, input_df, alpha):
+        """ BASED ON THE _select_kbest METHOD
+
+        Uses Scikit-learn's SelectFwe feature selection to filter the subset of features
+           according to p-values corresponding to Family-wise error rate
+        Parameters
+        ----------
+        input_df: pandas.DataFrame {n_samples, n_features+['class', 'group', 'guess']}
+            Input DataFrame to perform feature selection on
+        k: float, optional
```
Comment (rhiever) [inline tpot/tpot.py]: Please change this line to `alpha: float in the range [0.001, 0.05]`

File: tutorials/IRIS.ipynb [hunk 6]
```diff
@@ -87,28 +87,20 @@
   },
   {
    "cell_type": "code",
```
Comment (rhiever) [inline tutorials/IRIS.ipynb]: Please undo the changes to this tutorial.

File: tutorials/tpot_iris_pipeline.py [hunk 7]
```diff
@@ -2,28 +2,17 @@
 import pandas as pd
 
 from sklearn.cross_validation import StratifiedShuffleSplit
```
Comment (rhiever) [inline tutorials/tpot_iris_pipeline.py]: Please undo the changes to this file.

File: tpot/tpot.py [hunk 8]
```diff
@@ -348,8 +349,9 @@ def export(self, output_file_name):
 '''
         if '_variance_threshold' in operators_used: pipeline_text += 'from sklearn.feature_selection import VarianceThreshold\n'
         if '_select_kbest' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectKBest\n'
+        if '_select_fwe' in operators_used: pipeline_text += 'from sklearn.feature_selection import SelectFwe\n' # added selectfwe
```
Comment (rhiever) [inline tpot/tpot.py]: No need to have a comment saying `# added selectfwe` -- git and GitHub track those changes for us. :-)

File: tpot/tpot.py [hunk 9]
```diff
@@ -113,7 +113,8 @@ def __init__(self, population_size=100, generations=100,
         self.pset.addPrimitive(self.gradient_boosting, [pd.DataFrame, float, int, int], pd.DataFrame)
         self.pset.addPrimitive(self._combine_dfs, [pd.DataFrame, pd.DataFrame], pd.DataFrame)
         self.pset.addPrimitive(self._variance_threshold, [pd.DataFrame, float], pd.DataFrame)
-        self.pset.addPrimitive(self._select_kbest, [pd.DataFrame, int], pd.DataFrame) 
+        self.pset.addPrimitive(self._select_kbest, [pd.DataFrame, int], pd.DataFrame)
+        self.pset.addPrimitive(self._select_fwe, [pd.DataFrame, int], pd.DataFrame) # added selectfwe
```
Comment (rhiever) [inline tpot/tpot.py]: No need to have a comment saying `# added selectfwe` -- git and GitHub track those changes for us. :-)

**Comment by rhiever** [comment]
Thank you for this PR @kadarakos. This looks great! I commented on some of the lines of code to request some fixes before we merge it.

**Comment by kadarakos** [comment]
Thanks for the suggestions! I apologize for the messy PR, but I'm not very experienced in contributing to projects on github :-).

File: tpot/tpot.py [hunk 10]
```diff
@@ -113,7 +113,8 @@ def __init__(self, population_size=100, generations=100,
         self.pset.addPrimitive(self.gradient_boosting, [pd.DataFrame, float, int, int], pd.DataFrame)
         self.pset.addPrimitive(self._combine_dfs, [pd.DataFrame, pd.DataFrame], pd.DataFrame)
         self.pset.addPrimitive(self._variance_threshold, [pd.DataFrame, float], pd.DataFrame)
-        self.pset.addPrimitive(self._select_kbest, [pd.DataFrame, int], pd.DataFrame) 
+        self.pset.addPrimitive(self._select_kbest, [pd.DataFrame, int], pd.DataFrame)
+        self.pset.addPrimitive(self._select_fwe, [pd.DataFrame, int], pd.DataFrame)
```
Comment (rhiever) [inline tpot/tpot.py]: Just noticed -- this should be

```python
 self.pset.addPrimitive(self._select_fwe, [pd.DataFrame, float], pd.DataFrame)
```

since the second parameter that `select_fwe()` takes is the alpha value in the range [0.001, 0.05].

**Comment by rhiever** [comment]
No worries -- I appreciate that you took the time to contribute! :-) I just noted one more minor error, which when fixed, I believe will make this PR ready to merge.

**Comment by kadarakos** [comment]
Should I add anything else?

**Comment by rhiever** [comment]
Thank you for pinging me on this again! It's now merged. :-)
